### PR TITLE
Plugin refactor [1] : stricter data models, more explicit core logic 

### LINKED ISSUE #160: Problem with Keypoint Plugin
Hi, 

Thanks for the useful labelling gui. I made a standalone installation:
`pip install "pyside6==6.4.2"`
`pip install napari-deeplabcut`
`pip install "numpy<2.0.0"`
`pip install npe2==0.7.9` (this required because latest update on the library on Friday creates an error)

But the keypoint plugin does not working properly. Even if there is a plugin, I cannot change labeling mode and keypoint selection list is not available. 
Then I downgraded napari-deeplabcut to version 0.2.1.3 (I have seen some forum notes that it was a working version), after this downgrade the labeling mode was activated but keypoint selection list is still not visible. Could you please let me know what might be the problem and how I can solve it.. 

Thanks for your time.

### LINKED ISSUE #161: Old keypoints are erased when resuming labelling if config.yaml is loaded before labeled-data
Surprised this issue hasn't been reported before, but I went through every issue title on this repository and couldn't find an obvious match

If you partially label frames with napari, then save and close to come back later, the new labels added in a later session erase the old ones if you load config.yaml before loading the frames to be labelled. This is, to my mind, counter-intuitive behaviour that is not adequately explained in the UI and can cause catastrophic data loss without warning.

# Steps to reproduce

1. Download and unzip [this DLC project folder](https://github.com/user-attachments/files/25192796/TestNapari-John-2026-02-09.zip)
2. Open napari by typing `napari` on the command line (after having activated any necessary virtual environments)
3. Select Plugins->Keypoint controls (napari-deeplabcut) from the menu bar
4. Drag and drop the config.yaml from the attached folder into the central panel of the napari window
5. Drag and drop the folder labeled-data/test from the attached folder into the central panel of the napari window
6. Click somewhere to label bodypart1
7. Save by pressing Ctrl+S
8. Close napari
9. Repeat steps 2-5
10. Verify that bodypart1 is still labelled
11. Select bodypart2 from the Keypoint selection drop down on the rightmost panel
12. Click somewhere else to label bodypart 2
13. Repeat steps 7-9
 
# Expected behaviour

bodypart1 and bodypart2 are both labelled and both pairs of coordinates appear in labeled-data/test/CollectedData_John.csv

# Actual behaviour

Only bodypart2 is labelled when you open the third time and bodypart1's coordinates are erased from labeled-data/test/CollectedData_John.csv

# Additional info

While this behaviour is implied by the prompt that appears in the middle panel when loading the napari-deeplabcut plugin ("Drop a folder from within a DeepLabCut's labeled-data directory, and, *if labeling from scratch*, the corresponding project's config.yaml file" (emphasis mine)), I find this behaviour unsatisfactory, counter-intuitive, and inadequately sign-posted in a way that can cause users to lose data (indeed, already has in my case).

First, as a general user interaction principle, users should be warned and given the option to abort when operations have potentially catastrophic side effects (in this case, overwriting existing data). The most common comparison would be in most programs, the Save As... dialogue box warns you if you try to save a new file the same name as an old one.

Second, it is counter-intuitive that the way you open the data should affect how the data is saved. As I say, this is implied by the prompt, but understanding what is about to happen requires careful reading of an ambiguously worded prompt written in dark grey text on a black background. It's better not to rely on users reading things at all, but if you are going to make them read things, the behaviour should be spelled out more explicitly.

In my opinion, the default behaviour when labelling new keypoints/frames in an existing dataset should always be to add to that dataset, not overwrite the old keypoints. If the current behaviour is intended or there's some technical reason why it cannot be changed, it should be explicitly spelled out* and users should be warned when loading and/or saving (ideally both) a dataset in such a way that will cause old labels to be erased, and given the option to abort. If the current behaviour is going to remain, and there is a warning on saving, ideally the warning would give three options: overwrite (current behaviour), merge (old and new are both saved, my preferred default behaviour), cancel (nothing happens).

*i.e. the prompt should be something like "To label a dataset from scratch, drop the config.yaml from a DeepLabCut project and a folder from within its labeled-data directory (warning: this will erase any existing labels). To continue labeling a dataset, drop only a folder from within a DeepLabCut project's labeled-data directory".

# Environment info

OS: Windows 11 Enterprise version 24H2 build 26100.7462 experience 1000.26100.275.0
Python version: 3.10.14
conda version: 24.5.0
deeplabcut version: 3.0.0rc1
napari version: 0.4.18
napari-deeplabcut version: 0.2.1.7

Please let me know if you need any more info

### LINKED ISSUE #170: colorscheme not showing
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating System

macOS Sonoma Version 14.2.1

### DeepLabCut version

3.0.0rc13

### What engine are you using?

pytorch

### DeepLabCut mode

single animal

### Device type

Apple M1

### Bug description 🐛

I already added body parts when create the project but when Iabel frames in napari GUI, it didn't show anything in color scheme reference. When I added point, it only added hand_R (body parts I added are hand_R, hand_L, etc.) and don't know where to change to other body parts.

### Steps To Reproduce

_No response_

### Relevant log output

```shell

```

### Anything else?

_No response_

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by C-Achard**
## Core plugin refactor & maintainability improvements

Addresses the technical debt by gradually reducing legacy code, separating core logic from UI logic, and making data flow more centralized and explicit.

### TODO

- [x] Check if pydantic is >2.0 from napari on earlier python versions (3.10 esp.)
- [x] Check raising OSError in reader instead of returning None when no images are present

### Related

- Upstream of #164

---

###  Introduced `core/` foundation (UI/napari-free utilities)

A new internal `core/` package is now the canonical place for under-the-hood logic that can be shared by widgets, reader, and writer—without importing napari or Qt.

#### Added:

*   **`core/paths.py`**
    *   DLC path heuristics (detect config.yaml, detect labeled-data folders, detect DLC artifacts).
    *   Cross-platform path canonicalization (Windows/POSIX normalization).
    *   A `PathMatchPolicy` concept for deterministic matching behavior.

*   **`core/metadata.py`**
    *   Metadata inference utilities (e.g., infer root directory robustly).
    *   Safe merge/update/sync helpers (fill missing metadata fields).
    *   Uses pydantic models to help introduce validation in the future

*   **`core/remap.py`**
    *   Extracted frame/time remapping logic out of `_widgets.py`.
    *   Supports both array-like (Points/Tracks-like) and list-like (Shapes-like) data.
    *   Includes **ambiguity/risk detection** (duplicate canonical keys, low overlap/coverage, non-bijective mapping), emitting **warnings** (logging only—no UI behavior changes).
    *   Added unit tests for remapping behavior (including edge cases).

### 2) Added `config/` models with Pydantic

Metadata structure is now formalized using Pydantic models, while remaining compatible with napari layer metadata constraints.

> Pydantic is already required by napari so this does not add dependencies

#### Added:

*   **`config/models.py`**
    *   `ImageMetadata` and `PointsMetadata` models.
    *   A `DLCHeaderModel` wrapper (supports storing the runtime header object).
    *   Structured enums where useful.
    *   Clear invariants documented in docstrings.

> **Important design decision:** Models are authoritative internally, but napari layer metadata remains dict-like and is updated in-place to avoid breaking napari behavior. May be changed in the future.

### 3) `_widgets.py` cleaned up and made more maintainable

The main widget codebase was substantially de-risked by removing brittle hooks and reducing private API reliance.

#### Notable improvements:

*   **Removed the “Welcome overlay” hack** that was intercepting drag-and-drop in newer napari versions.
    *   Fixes the bug: *drag/drop works only when the widget is closed*.
    *   Removes a fragile dependency on napari internal widget wiring.

*   **Refactored frame remapping into core**
    *   `_widgets.py` now delegates to `core.remap.remap_layer_data_by_paths`.
    *   Widget code now contains only the napari glue (extract paths, determine time column, set layer.data).

*   **Unified canonical path usage**
    *   Prefer `core.paths.canonicalize_path` instead of scattered/legacy patterns.
    *   Path policy for overlap resolution moved out of UI logic and into core.

### Added unit tests for `core/remap.py`

*   New tests under `src/napari_deeplabcut/_tests/core/` validate:
    *   index-map construction
    *   depth fallback behavior
    *   array-like remapping and list-like remapping
    *   no-op cases
    *   unmapped values remain unchanged
    *   ambiguous/risky conditions emit warnings (logging-based)

### New E2E and widget tests

*   Tests were adjusted to remain stable under the new “authoritative model + napari metadata dict” boundary.
*   Modal dialogs remain patched in tests to avoid hangs.

### Drag-and-drop no-op when widget is open

**Cause:** the replaced Welcome overlay widget swallowed drop events due to changes in napari’s internal overlay wiring (private API drift).  
**Fix:** removed the overlay replacement entirely and returned drag/drop responsibility to napari. The welcome menu was not really used and we have a custom tutorial in place already for user guidance.

***

## Architectural Direction Established

This PR establishes the new architecture baseline:

*   **`core/`**: deterministic logic, shared, napari-free, testable
*   **`config/`**: Pydantic models/enums/schema definitions
*   **widgets**: UI + napari interactions + orchestration only
*   **no “services/” layer**: logic belongs in `core/`

This provides a stable foundation for the next refactor phases (dataframe pipeline unification, reader/writer consolidation, and metadata consistency enforcement).

***

### To be addressed in subsequent PRs:

*   DataFrame transformations are still split across `_reader.py`, `_writer.py`, and widget code paths.
*   Path canonicalization is not fully unified across reader/writer yet (reader still uses some `misc` utilities).
*   Some legacy paths and metadata keys remain for backward compatibility.
*   The `Show trails` multi-animal NaN crash is identified but not yet fixed in this PR (planned next).

***

## Next Steps (Planned)

1.  **Fix “Show trails” crash for multi-animal** by filtering non-finite points before creating Tracks.
  - This was fixed in #114 but needs to be adapted for this refactor
3.  **DataFrame/pandas pipeline consolidation**:
    *   Create a `core/dataframe.py` to unify:
        *   Points → DLC DataFrame
        *   DLC DataFrame → Points data/properties
        * Consistent index/path rules
        * Ensure ease of refactoring for the upcoming DLC refactor
4.  Reduce duplication between `_reader.py` and `_writer.py` and enforce a single metadata schema pathway using newly introduced models
5.  Continue thinning `_widgets.py` and optionally split UI into smaller modules once logic is fully extracted.

### CODE REVIEW

File: src/napari_deeplabcut/_widgets.py [hunk 1]
```diff
@@ -55,16 +55,24 @@
...
-    canonicalize_path,
     encode_categories,
     guarantee_multiindex_rows,
-    remap_array,
 )
 
 Tip = namedtuple("Tip", ["msg", "pos"])
 
+logger = logging.getLogger("napari-deeplabcut._widgets")
+logger.setLevel(logging.DEBUG)
```
[the remark this answers is not in the document]
Comment (C-Achard) [reply src/napari_deeplabcut/_widgets.py]: Leaving open as reminder

File: src/napari_deeplabcut/_widgets.py [hunk 2]
```diff
@@ -561,25 +552,19 @@ def __init__(self, napari_viewer):
         self.viewer.layers.events.inserted.connect(self.on_insert)
         self.viewer.layers.events.removed.connect(self.on_remove)
 
-        self.viewer.window.qt_viewer._get_and_try_preferred_reader = MethodType(
-            _get_and_try_preferred_reader,
-            self.viewer.window.qt_viewer,
-        )
+        # self.viewer.window.qt_viewer._get_and_try_preferred_reader = MethodType(
```
Comment (C-Achard) [inline src/napari_deeplabcut/_widgets.py:555]: Was unsafe, we should let napari handle things in a best effort way as it already does

File: src/napari_deeplabcut/_widgets.py [hunk 3]
```diff
@@ -561,25 +552,19 @@ def __init__(self, napari_viewer):
...
         status_bar.addPermanentWidget(self.last_saved_label)
 
-        # Hack napari's Welcome overlay to show more relevant instructions
-        overlay = self.viewer.window._qt_viewer._welcome_widget
-        welcome_widget = overlay.layout().itemAt(1).widget()
-        welcome_widget.deleteLater()
-        w = QtWelcomeWidget(None)
-        overlay._overlay = w
-        overlay.addWidget(w)
-        # overlay._overlay.sig_dropped.connect(overlay.sig_dropped)
```
Comment (C-Achard) [inline src/napari_deeplabcut/_widgets.py:581]: This API was no longer valid, and overriding the  drag&drop behavior

File: src/napari_deeplabcut/_widgets.py [hunk 4]
```diff
@@ -697,6 +682,154 @@ def __init__(self, napari_viewer):
...
+        # After adoption, refresh UI state
+        try:
+            active = self.viewer.layers.selection.active
+            if active is not None:
+                # Force the GUI to update visibility of menus, etc.
+                self.on_active_layer_change(Event(type_name="active", value=active))
+        except Exception:
+            pass
+
+    def _adopt_layer(self, layer, index: int) -> None:
```
Comment (C-Achard) [inline src/napari_deeplabcut/_widgets.py:708]: Note : duplicated logic from on_insert. Should be separated and moved to e.g. a layers.py file

File: src/napari_deeplabcut/_widgets.py [hunk 5]
```diff
@@ -1550,116 +1695,6 @@ def item_changed(ind):
...
-        Provide style updates on event.
-
-        Parameters
-        ----------
-        event : qtpy.QtCore.QEvent
-            Event from the Qt context.
-        """
-        self._update_property("drag", False)
-
-    def dropEvent(self, event):
```
Comment (C-Achard) [inline src/napari_deeplabcut/_widgets.py:1649]: This was the problematic override for drag&drop

File: src/napari_deeplabcut/_writer.py [hunk 6]
```diff
@@ -44,41 +52,137 @@ def _form_df(points_data, metadata):
...
-    name = metadata["name"]
-    root = meta["root"]
-    if "machine" in name:  # We are attempting to save refined model predictions
-        df.drop("likelihood", axis=1, level="coords", inplace=True, errors="ignore")
-        header = misc.DLCHeader(df.columns)
+    """
+    Write DLC keypoints to disk.
+
+    SAFETY POLICY:
+    - For ground-truth files (CollectedData_*), never delete existing labels silently.
```
Comment (C-Achard) [inline src/napari_deeplabcut/_writer.py:59]: Note : this is currently not fully implemented/used throughout.
We also likely want it to be configurable ("skip overwrite confirmation")

File: src/napari_deeplabcut/config/models.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+
+# -----------------------------------------------------------------------------
+# Header model (authoritative wrapper)
+# -----------------------------------------------------------------------------
+
+
+class DLCHeaderModel(BaseModel):
+    """
+    Structured representation of a DLC header.
```
Comment (C-Achard) [inline src/napari_deeplabcut/config/models.py:27]: Note : this should be flexible for when DLC refactor hits

File: src/napari_deeplabcut/config/models.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
+# src/napari_deeplabcut/config/models.py
+from __future__ import annotations
+
+from enum import Enum
+from typing import Any
+
+from pydantic import BaseModel, ConfigDict, Field
```
Comment (C-Achard) [inline src/napari_deeplabcut/config/models.py:7]: No validations yet, worth adding later

File: src/napari_deeplabcut/core/paths.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+    if not p.exists() or not p.is_dir():
+        return False
+
+    if has_dlc_datafiles(p):
+        return True
+
+    return any(part.lower() == "labeled-data" for part in p.parts)
+
+
+def should_force_dlc_reader(paths: str | Path | Iterable[str | Path]) -> bool:
```
Comment (C-Achard) [inline src/napari_deeplabcut/core/paths.py:158]: Worth reviewing whether it should "force" anything in the first place, especially with latest plugin reader changes in napari

File: src/napari_deeplabcut/core/remap.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+from dataclasses import dataclass
+from typing import Any
+
+import numpy as np
+
+from napari_deeplabcut.core.paths import PathMatchPolicy, canonicalize_path, find_matching_depth
+
+logger = logging.getLogger(__name__)
+
+# Heuristic thresholds for "risky remap" warnings.
```
Comment (C-Achard) [inline src/napari_deeplabcut/core/remap.py:15]: Given the relatively short paths this may not be the best approach, but it might help debugging nonetheless
